File: backend/src/main/java/com/laptopdb/backend/entity/Gpu.java
package com.laptopdb.backend.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import lombok.Data;

@Data
@Entity
@Table(name = "gpus")
public class Gpu {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @NotBlank(message = "Brand cannot be blank")
    @Column(nullable = false, length = 50)
    private String brand;

    // ENGINEERING DECISION: UNIQUE constraint omitted to allow variants with different TDPs.
    @NotBlank(message = "Model name cannot be blank")
    @Column(name = "model_name", nullable = false, length = 100)
    private String modelName;

    // NULLABLE FIELDS: Can be left null for integrated graphics (e.g., Iris Xe).
    @Column(name = "tdp_watt")
    private Integer tdpWatt;

    @Column(name = "vram_gb")
    private Integer vramGb;

    @Column(name = "vram_type", length = 20)
    private String vramType;

    @Column(name = "memory_bus_bit")
    private Integer memoryBusBit;
}

File: frontend/lib/models/gpu.dart
class Gpu {
  final int id;
  final String brand;
  final String modelName;
  final int? tdpWatt;
  final int? vramGb;
  final String? vramType;
  final int? memoryBusBit;

  Gpu({
    required this.id,
    required this.brand,
    required this.modelName,
    this.tdpWatt,
    this.vramGb,
    this.vramType,
    this.memoryBusBit,
  });

  factory Gpu.fromJson(Map<String, dynamic> json) {
    return Gpu(
      id: json['id'] ?? 0,
      brand: json['brand'] ?? '',
      modelName: json['modelName'] ?? '',
      tdpWatt: json['tdpWatt'],
      vramGb: json['vramGb'],
      vramType: json['vramType'],
      memoryBusBit: json['memoryBusBit'],
    );
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'brand': brand,
      'modelName': modelName,
    };
    if (tdpWatt != null) map['tdpWatt'] = tdpWatt;
    if (vramGb != null) map['vramGb'] = vramGb;
    if (vramType != null) map['vramType'] = vramType;
    if (memoryBusBit != null) map['memoryBusBit'] = memoryBusBit;
    return map;
  }

  String get displayLabel => '$brand $modelName${vramGb != null ? ' ${vramGb}GB' : ''}';
}
